fenix/app/src/main/java/org/mozilla/fenix/FenixApplication.kt

```diff
@@ -521,13 +524,61 @@ open class FenixApplication : LocaleAwareApplication(), Provider {
      * thread, early in the app startup sequence.
      */
     private fun beginSetupMegazord() {
-        // Note: Megazord.init() must be called as soon as possible ...
-        Megazord.init()
+        try {
+            // Note: Megazord.init() must be called as soon as possible ...
+            Megazord.init()
+
+            initializeRustErrors(components.analytics.crashReporter)
+            // ... but RustHttpConfig.setClient() and RustLog.enable() can be called later.
 
-        initializeRustErrors(components.analytics.crashReporter)
-        // ... but RustHttpConfig.setClient() and RustLog.enable() can be called later.
+            RustLog.enable()
+        } catch (e: UnsatisfiedLinkError) {
+            @Suppress("TooGenericExceptionCaught")
+            try {
+                reportUnsatisfiedLinkErrorBreadcrumbs()
+            } catch (e: Throwable) {
+                // This shouldn't happen, but if it does it's better to ignore the exception from
+                // the breadcrumb code and rethrow the initial exception.
+            }
+            throw e
+        }
+    }
 
-        RustLog.enable()
+    private fun reportUnsatisfiedLinkErrorBreadcrumbs() {
+        val breadcrumbs = mutableListOf<String>()
+        val apkPath = applicationContext.getApplicationInfo().sourceDir
+        breadcrumbs.add("APK: $apkPath")
+        val apkDir = File(apkPath).getParentFile()
+        val installSourcePackage = if (SDK_INT >= Build.VERSION_CODES.R) {
+            packageManager.getInstallSourceInfo(packageName).installingPackageName
+        } else {
+            @Suppress("DEPRECATION")
+            packageManager.getInstallerPackageName(packageName)
+        }
+        breadcrumbs.add("Installer package name: $installSourcePackage")
+
+        val apkDirFileList = apkDir?.let {
+            it.walk()
+                .filter { it != apkDir }
+                .map { it.relativeTo(apkDir).toString() }
+                .joinToString(", ")
+        }
+        breadcrumbs.add("Files inside dir: $apkDirFileList")
+
+        val apkFileList = ZipInputStream(FileInputStream(apkPath)).use {
+            generateSequence { it.nextEntry }.joinToString(", ")
+        }
```

Comment:
A little slower when we're about to crash is acceptable in my opinion.  Is there a way for us to filter those files so we can turn the file list into a list of true/false? (x file exists, y file exist etc)

Reply:
I tested it my emulator (Pixel 6, x86-64).

It would definitely be possible to transform the file list, but maybe it should more like a diff:  X files were present in the APK, but not on the device -- Y files were present on the device, but not the APK.